src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,330 @@
+  * separate from the main State business.
+  */
+trait ZKBackgroundPromiseSupport { self: ZookeeperDirectState =>
+
+  /**
+    * A Callable implement that is not supposed to be called.
+    */
+  class Uncallable[A] extends Callable[A] {
+    override def call(): A = throw new IllegalArgumentException("The call() method is unsupported")
+  }
+
+  /**
+    * An inoperable Java Future. Typically used to contain a Scala Future using ScalaFutureSupport.
+    */
+  class UncallableJavaFuture[A] extends FutureTask[A](new Uncallable) {
+    override def get(l: Long, timeUnit: java.util.concurrent.TimeUnit): A = get()
+    override def get(): A = throw new IllegalArgumentException("get() is unsupported - call future() instead")
+  }
+
+  class ScalaFutureWrapper[A](val future: SFuture[A])
+    extends UncallableJavaFuture[A]
+    with ScalaFutureSupport[A]
+
+  implicit class ToScalaFutureWrapper[A](val future: SFuture[A]) {
+    def asScalaFutureWrapper = new ScalaFutureWrapper(future)
+  }
+
+  /**
+    * Base class for our combination Curator callbacks, Scala promises, Scala future containers
+    * and Java futures... all in one!
+    */
+  abstract class CallbackPromise[A] extends BackgroundCallback {
+    val prommy: Promise[A] = Promise[A]()
+    def future: SFuture[A] = prommy.future
+  }
+
+  /**
+    * A callback for zookeeper.getData() which returns it wrapped in a Future.
+    *
+    * Uses a Promise to generate a Scala Future. Also extends Java Future, but will throw an exception if
+    * called as a Java Future. This allows us to pass the Scala Future directly to Marathon/Chronos
+    * via the Java-Future-based State interface.
+    */
+  class GetDataCallbackPromise(name: String) extends CallbackPromise[Variable] {
+
+    override def processResult(client: CuratorFramework, event: CuratorEvent): Unit = {
+      val entry = StateProtos.Entry.parseFrom(event.getData)
+      val lv = new LocalVariable(name, entry.getValue.toByteArray, Some(entry.getUuid.toStringUtf8))
+      prommy.success(lv)
+    }
+
+    def success(v: Variable): Unit = { prommy.success(v) }
+  }
+
+  /**
+    * A callback for zookeeper.setData() which then fetches the variable and returns it wrapped in a Future.
+    * In other words, uses EventualVariableFetch after setData() calls this back.
+    *
+    * Uses a Promise to generate a Scala Future. Also extends Java Future, but will throw an exception if
+    * called as a Java Future. This allows us to pass the Scala Future directly to Marathon/Chronos
+    * via the Java-Future-based State interface.
+    */
+  class SetDataCallbackPromise(name: String) extends CallbackPromise[Variable] {
+
+    private val eventualFetch = new GetDataCallbackPromise(name)
+
+    override def future: SFuture[Variable] = eventualFetch.future
+
+    override def processResult(client: CuratorFramework, event: CuratorEvent): Unit = {
+      handleFetch(name, eventualFetch)
```

Comment:
Why fetch the variable again? Make contain a `Future[Unit]` and in the handler do `callback.future.map(_ => variable)`, because we already have the variable we are trying to store ;-).